### PR TITLE
feat(core): wrap @logto/shared and @logto/connector-types

### PR DESCRIPTION
**Pull request by darcyYe**
## Summary

1. create @logto/shared package: contains logto shared util functions
2. create @logto/connector-types package: connector basic types definition

## Testing

Passed UTs.

### CODE REVIEW

File: packages/connector-types/jest.config.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import type { Config } from '@jest/types';
```
Comment (gao-sun) [inline packages/connector-types/jest.config.ts:1]: @darcyYe @simeng-li @wangsijie we need a shared jest config i think. mind to add an issue?

File: packages/connector-types/src/aliyun.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+      return value === undefined ? '' : `${escaper(key)}=${escaper(value)}`;
+    })
+    .filter(Boolean)
+    .join('&');
+  const stringToSign = `${method.toUpperCase()}&${escaper('/')}&${escaper(canonicalizedQuery)}`;
+
+  return createHmac('sha1', `${secret}&`).update(stringToSign).digest('base64');
+};
+export interface PublicParameters {
```
Comment (gao-sun) [inline packages/connector-types/src/aliyun.ts:39]:
```suggestion
};

export interface PublicParameters {
```

File: packages/connector-types/src/aliyun.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
+import { createHmac } from 'crypto';
```
Comment (gao-sun) [inline packages/connector-types/src/aliyun.ts:1]: is this file required over aliyun sms/email and social? will it be better to specify what would import this

File: packages/connector-types/package.json [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    "lint-staged": "^11.1.1",
+    "prettier": "^2.3.2",
+    "ts-jest": "^27.1.1",
+    "typescript": "^4.6.2"
+  },
+  "eslintConfig": {
+    "extends": "@silverhand",
+    "rules": {
+      "@typescript-eslint/ban-types": "off"
+    }
```
Comment (gao-sun) [inline packages/connector-types/package.json:49]: why do we need ban this?

File: packages/connector-types/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+type SmsMessageTypes = EmailMessageTypes;
+
+export type SendEmailResponse = { EnvId: string; RequestId: string };
+
+export type SendSmsResponse = { BizId: string; Code: string; Message: string; RequestId: string };
+
+export type EmailSendMessageFunction<
+  T1 extends ArbitraryObject = ArbitraryObject,
+  T2 = Response<T1>
+> = (
```
Comment (gao-sun) [inline packages/connector-types/src/index.ts:63]: why do we have aliyun response as the default value?

File: packages/connector-types/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+};
+
+type SmsMessageTypes = EmailMessageTypes;
+
+export type SendEmailResponse = { EnvId: string; RequestId: string };
+
+export type SendSmsResponse = { BizId: string; Code: string; Message: string; RequestId: string };
+
+export type EmailSendMessageFunction<
+  T1 extends ArbitraryObject = ArbitraryObject,
```
Comment (gao-sun) [inline packages/connector-types/src/index.ts:61]: will it be better to use meaningful names in generic when it has more than one type?

File: packages/connector-types/tsconfig.base.json [hunk 7]
```diff
@@ -0,0 +1,10 @@
+{
+  "extends": "@silverhand/ts-config/tsconfig.base",
+  "compilerOptions": {
+    "outDir": "build",
```
Comment (gao-sun) [inline packages/connector-types/tsconfig.base.json:4]: this is more like a lib, so will it be better to keep the default option "lib" here?

File: packages/connector-utils/package.json [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+  "scripts": {
+    "preinstall": "npx only-allow pnpm",
+    "precommit": "lint-staged",
+    "build": "rm -rf build/ && tsc --p tsconfig.build.json",
+    "lint": "eslint --ext .ts src",
+    "lint:report": "pnpm lint -- --format json --output-file report.json",
+    "prepack": "pnpm build"
+  },
+  "engines": {
+    "node": ">=14.15.0"
```
Comment (gao-sun) [inline packages/connector-utils/package.json:20]: we use `^16.0.0` now

File: packages/connector-utils/package.json [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    "eslint": "^8.10.0",
+    "lint-staged": "^11.1.1",
+    "prettier": "^2.3.2",
+    "typescript": "^4.6.2"
+  },
+  "eslintConfig": {
+    "extends": "@silverhand",
+    "rules": {
+      "@typescript-eslint/ban-types": "off"
+    }
```
Comment (gao-sun) [inline packages/connector-utils/package.json:36]: ditto

File: packages/connector-utils/tsconfig.base.json [hunk 10]
```diff
@@ -0,0 +1,10 @@
+{
+  "extends": "@silverhand/ts-config/tsconfig.base",
+  "compilerOptions": {
+    "outDir": "build",
```
Comment (gao-sun) [inline packages/connector-utils/tsconfig.base.json:4]: this should match `files` in `package.json`

File: packages/connector-utils/package.json [hunk 11]
```diff
@@ -0,0 +1,39 @@
+{
```
Comment (gao-sun) [inline packages/connector-utils/package.json:1]: forgot to change directory name?

File: packages/connector-types/jest.config.ts (hunk 1 above)
Comment (darcyYe) [reply packages/connector-types/jest.config.ts:1]: [LOG-2271](https://linear.app/silverhand/issue/LOG-2271/add-a-shared-jest-config) was added to track this.

Comment (gao-sun) [reply packages/connector-types/jest.config.ts:1]: @darcyYe cool, please see #644

File: packages/connector-types/src/aliyun.ts (hunk 3 above)
Comment (darcyYe) [reply packages/connector-types/src/aliyun.ts:1]: Should reorganize these constants and methods in aliyun.ts.
They are initially added for aliyun-dm and aliyun-sms connectors and are not general enough. We may simply put code in aliyun.ts in connector-aliyun-sms and connector-aliyun-dm, or wrap them as a new package (such as @logto/aliyun-utils).
At this stage, I personally prefer the first solution as we do not have many aliyun-based services.

File: packages/connector-types/src/index.ts (hunk 5 above)
Comment (darcyYe) [reply packages/connector-types/src/index.ts:63]: This is actually Response from got. As EmailSendMessageFunction/SmsSendMessageFunction are wrapped HTTP API call, the return type is expected to be got response.

File: packages/connector-types/jest.config.ts (hunk 1 above)
Comment (simeng-li) [reply packages/connector-types/jest.config.ts:1]: good suggestion, BTW @darcyYe  do we need to process `'^jose/(.*)$': '<rootDir>/node_modules/jose/dist/node/cjs/$1',` for connector package?  Does it include `jose` as its dependency?

File: packages/connector-types/src/aliyun.test.ts [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+      RegionId: 'cn-hangzhou',
+      ReplyToAddress: 'true',
+      SignatureMethod: 'HMAC-SHA1',
+      SignatureNonce: 'c1b2c332-4cfb-4a0f-b8cc-ebe622aa0a5c',
+      SignatureVersion: '1.0',
+      Subject: '3',
+      TagName: '2',
+      Timestamp: '2016-10-20T06:27:56Z',
+      ToAddress: '1@test.com',
+      Version: '2015-11-23',
```
Comment (simeng-li) [inline packages/connector-types/src/aliyun.test.ts:24]: could extract to a common mock variable?

File: packages/connector-types/src/aliyun.ts [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+  encodeURIComponent(string_)
+    .replace(/\*/g, '%2A')
+    .replace(/'/g, '%27')
+    .replace(/!/g, '%21')
+    .replace(/"/g, '%22')
+    .replace(/\(/g, '%28')
+    .replace(/\)/g, '%29')
+    .replace(/\+/, '%2B');
+
+export const getSignature = (
```
Comment (simeng-li) [inline packages/connector-types/src/aliyun.ts:20]: Is this supposed to be a util methods file?

File: packages/shared/src/index.ts [hunk 14]
```diff
@@ -0,0 +1,9 @@
+import { existsSync, readFileSync } from 'fs';
+
+export const getMarkdownContents = (filePath: string, fallbackContent: string): string => {
```
Comment (simeng-li) [inline packages/shared/src/index.ts:3]: since this is a shared methos,  tell from the function content itself, not necessarily  MarkDown contents?

File: packages/connector-types/package.json [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+    "test:coverage": "jest --coverage --silent"
+  },
+  "engines": {
+    "node": "^16.0.0"
+  },
+  "dependencies": {
+    "@logto/phrases": "^0.1.0",
+    "@logto/schemas": "^0.1.0",
+    "got": "^11.8.2",
+    "jose": "^3.14.3"
```
Comment (simeng-li) [inline packages/connector-types/package.json:28]: ditto: do we need jose here?

File: packages/connector-types/src/aliyun.ts (hunk 3 above)
Comment (wangsijie) [reply packages/connector-types/src/aliyun.ts:1]: Agree with first solution.

Is not too late to refactor when we have 3rd aliyun's connector.

File: packages/connector-types/src/index.ts [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+  id: string;
+  type: ConnectorType;
+  name: Record<Languages, string>;
+  logo: string;
+  description: Record<Languages, string>;
+  readme: string;
+  configTemplate: string;
+}
+
+export interface ConnectorDTO extends Connector {
```
Comment (wangsijie) [inline packages/connector-types/src/index.ts:18]: This belongs to `core`.

File: packages/connector-types/package.json [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+    "lint": "eslint --ext .ts src",
+    "lint:report": "pnpm lint -- --format json --output-file report.json",
+    "prepack": "pnpm build"
+  },
+  "engines": {
+    "node": "^16.0.0"
+  },
+  "dependencies": {
+    "@logto/phrases": "^0.1.0",
+    "@logto/schemas": "^0.1.0",
```
Comment (wangsijie) [inline packages/connector-types/package.json:24]: `connector-types` should not have this dependency.

**Review by wangsijie — CHANGES_REQUESTED** [review]

File: packages/shared/tsconfig.base.json [hunk 18]
```diff
@@ -0,0 +1,10 @@
+{
+  "extends": "@silverhand/ts-config/tsconfig.base",
+  "compilerOptions": {
+    "outDir": "lib",
+    "baseUrl": ".",
+    "paths": {
+      "@/*": ["src/*"]
+    }
+  }
+}
```
Comment (gao-sun) [inline packages/shared/tsconfig.base.json:10]: this can be merged into `tsconfig.json`

File: packages/shared/tsconfig.build.json [hunk 19]
```diff
@@ -0,0 +1,5 @@
+{
+  "extends": "./tsconfig",
+  "include": ["src"],
```
Comment (gao-sun) [inline packages/shared/tsconfig.build.json:3]: we already have `include` from extended config, no need for this line?

File: packages/connector-types/tsconfig.base.json [hunk 20]
```diff
@@ -0,0 +1,10 @@
+{
+  "extends": "@silverhand/ts-config/tsconfig.base",
```
Comment (gao-sun) [inline packages/connector-types/tsconfig.base.json:2]: ditto

File: packages/connector-types/tsconfig.build.json [hunk 21]
```diff
@@ -0,0 +1,5 @@
+{
+  "extends": "./tsconfig",
+  "include": ["src"],
```
Comment (gao-sun) [inline packages/connector-types/tsconfig.build.json:3]: ditto

File: packages/connector-types/src/index.ts [hunk 22]
```diff
@@ -0,0 +1,107 @@
+import { Languages } from '@logto/phrases';
```
Comment (gao-sun) [inline packages/connector-types/src/index.ts:1]:
```suggestion
import type { Languages } from '@logto/phrases';
```

this can just import type from the package

**Review by gao-sun — APPROVED** [review]
overall looks great tough

File: packages/shared/src/index.ts (hunk 14 above)
Comment (darcyYe) [reply packages/shared/src/index.ts]: changed to `getFileContents`

**Review by wangsijie — APPROVED** [review]
